packages/world/src/utils/InventoryUtils.sol

```diff
@@ -317,145 +380,188 @@ library InventoryUtils {
-    } else {
-      slot = Inventory._getNextSlot(owner);
-      uint16 maxSlots = owner.getObjectType().getMaxInventorySlots();
-      require(slot < maxSlots, "Inventory is full");
-      Inventory._setNextSlot(owner, slot + 1);
+    for (uint256 i = 0; i < bitmapLength; i++) {
+      uint256 word = InventoryBitmap._getItem(owner, i);
+      while (word != 0) {
+        count++;
+        word &= word - 1;
+      }
     }
-
-    _addToOccupiedSlots(owner, slot);
   }
 
-  function _useEmptySlot(EntityId owner, uint16 slot) private {
-    uint16 maxSlots = owner.getObjectType().getMaxInventorySlots();
-    require(slot < maxSlots, "Invalid slot");
+  function findObjectType(EntityId owner, ObjectType objectType) internal view returns (uint16 slot) {
+    uint256 bitmapLength = InventoryBitmap._length(owner);
 
-    InventorySlotData memory slotData = InventorySlot._get(owner, slot);
-    require(slotData.objectType.isNull(), "Slot must be empty");
+    for (uint256 i = 0; i < bitmapLength; i++) {
+      uint256 word = InventoryBitmap._getItem(owner, i);
+      while (word != 0) {
+        uint256 bitIndex = LibBit.ffs(word);
+        word &= word - 1;
 
-    _addToOccupiedSlots(owner, slot);
+        slot = uint16(i * SLOTS_PER_WORD + bitIndex);
+        if (objectType == InventorySlot._getObjectType(owner, slot)) {
+          return slot;
+        }
+      }
+    }
 
-    uint16 nextSlot = Inventory._getNextSlot(owner);
+    revert("Object type not found");
+  }
 
-    // if slot < nextSlot, this slot is already tracked
-    if (slot < nextSlot) {
-      _removeFromTypeSlots(owner, ObjectTypes.Null, slot);
-      return;
-    }
+  function findEntity(EntityId owner, EntityId entityId) internal view returns (uint16 slot) {
+    uint256 bitmapLength = InventoryBitmap._length(owner);
+
+    for (uint256 i = 0; i < bitmapLength; i++) {
+      uint256 word = InventoryBitmap._getItem(owner, i);
+      while (word != 0) {
+        uint256 bitIndex = LibBit.ffs(word);
+        word &= word - 1;
 
-    // Fill gaps in the null type slots
-    for (uint16 i = nextSlot; i < slot; i++) {
-      _addToTypeSlots(owner, ObjectTypes.Null, i);
+        slot = uint16(i * SLOTS_PER_WORD + bitIndex);
+        if (entityId == InventorySlot._getEntityId(owner, slot)) {
+          return slot;
+        }
+      }
     }
 
-    Inventory._setNextSlot(owner, slot + 1);
+    revert("Entity not found");
   }
 
-  // Marks a slot as empty - O(1)
-  function _recycleSlot(EntityId owner, uint16 slot) private {
-    ObjectType objectType = InventorySlot._getObjectType(owner, slot);
+  /// @dev Removes empty words from the end of the bitmap.
+  function _pruneTrailingEmptyWords(EntityId owner) private {
+    uint256 length = InventoryBitmap._length(owner);
+    while (length != 0) {
+      uint256 lastWord = InventoryBitmap._getItem(owner, length - 1);
+      if (lastWord != 0) break;
+      InventoryBitmap._popBitmap(owner);
+      unchecked {
+        --length;
```

Comment:
out of curiosity, is this cheaper gas wise because it doesnt do the underflow check (and similarly overflow check in `_ensureWordExists`?

Reply:
yes, we could do it everywhere but it is just ugly, here I felt it was fine